[3 - 4단계 방탈출 예약 대기] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->


테스트 대역을 사용하여 service 클래스를 테스트하다 보니, 서비스에 큰 기능이 없었고 데이터베이스와 관련된 저장, 삭제, 수정, 조회의 경우에도 다 하나의 기능에 포함되어 있는 것 이라고 판단하여 h2를 fake라고 생각하고 테스트를 작성해 보았습니다.
- h2를 fake로 사용할 수 있다고 생각했던 근거는, mysql, postgresql 같은 데이터베이스와는 다르게 인메모리에서 동작하기 때문에 비교적 속도가 빠르고, 테스트와 함께 실행되기 때문에 환경에 대한 통제도 충분히 할 수 있을 것이라 판단하였습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

@coderabbitai full review

changuii committed:
refactor: 생성자의 접근 제어자를 public에서 protected로 변경

- 엔티티 클래스(ReservationTime, Theme, Member, Reservation)의 기본 생성자 접근 제어자를 protected로 변경하여 불필요한 외부 사용을 제한
- Member 클래스에서 reservations 필드 초기화를 constructor 내에서 직접 처리

changuii committed:
style: Boolean 타입을 boolean 타입으로 변경

changuii committed:
refactor: 예외 처리 및 코드 리팩토링

- 람다 표현식을 메서드 참조로 변경하여 코드 가독성 개선
- 불필요한 변수 제거 및 메서드 반환 구조 간소화
- 상수를 static final로 변경하여 일관성 확보
- JWT 서명 방식 변경에 따른 메서드 수정

changuii committed:
refactor: MemberRepository를 MemberRepositoryFacade로 추상화하여 서비스 계층 의존성 변경

- MemberRepositoryFacade 및 구현체(MemberRepositoryFacadeImpl) 생성
- MemberService에서 MemberRepository 대신 MemberRepositoryFacade 사용으로 변경
- Reservation에 연관관계 헬퍼 메서드 추가 (makeBiDirectionAssociation)
- JwtProvider의 JWT 파싱 로직 오타 수정 (parseClaimsJwt -> parseClaimsJws)

changuii committed:
test: MemberServiceTest 리팩토링 및 auto_increment 초기화 sql 추가

- MemberRepository를 MemberRepositoryFacade로 변경하여 spy를 사용하도록 수정
- 테스트 데이터 생성을 위한 SQL 스크립트 추가 및 적용 (@Sql 사용)
- assertSoftly를 활용하여 테스트 검증 로직 개선
- 불필요한 mock 및 stubbing 제거 후 테스트 로직 간결화
- 관련 repository 주입 및 초기화 작업 추가

changuii committed:
test: Member의 비밀번호 검증 테스트 및 관련 메서드 삭제

changuii committed:
test: MemberServiceTest 설정 수정 및 불필요한 setup 메서드 제거

changuii committed:
`chore: 초기화 스크립트 파일명 변경 및 테이블 데이터 리셋 쿼리 추가`

changuii committed:
refactor(theme): ThemeRepositoryFacade 도입 및 서비스 레이어 리팩토링

ThemeRepositoryFacade를 추가하여 ThemeRepository에 대한 접근을 캡슐화하고, 이를 통해 ThemeService의 의존성을 ThemeRepository에서 ThemeRepositoryFacade로 교체하여 코드의 유연성과 테스트 용이성을 개선했습니다. 관련 테스트 코드도 이에 맞게 수정하였습니다.

changuii committed:
refactor(reservationtime): ReservationTimeRepository 의존성을 Facade로 변경하여 유연성 향상

ReservationTimeRepository를 ReservationTimeRepositoryFacade로 추상화하고 관련된 구현체(ReservationTimeRepositoryFacadeImpl)를 추가. 서비스 계층에서 repository 접근 시 Facade를 사용하도록 수정하여 코드 의존성을 줄이고 확장성을 강화.

changuii committed:
test: RepositoryFacade 구현체 적용 및 테스트 코드 리팩토링

테스트 코드에서 기존의 개별 Repository를 RepositoryFacade 구현체로 변경하고 이에 따른 의존성을 수정함.
중복 코드 제거 및 테스트 작성 방식을 개선하여 가독성과 유지보수성을 향상시킴.

changuii committed:
refactor(reservation): 리팩토링 및 repository facade 패턴 적용

Member 및 Reservation 관련 로직을 repository facade 패턴으로 재구성하여 코드 구조를 개선하였습니다.
- `ReservationRepositoryFacade` 추가 및 구현
- Member와 Reservation 간 양방향 연관관계를 단방향으로 변경
- DTO 및 메서드 네이밍 개선
- 테스트 코드에서 Mock 사용 제거 및 SpyBean으로 변경
- `/members/reservations` API를 `/reservations/mine`으로 리팩토링

changuii committed:
test: AuthServiceTest 리팩토링 및 StubJwtProvider 삭제

- @DataJpaTest 및 @Sql 사용하여 테스트 환경 구성 개선
- StubJwtProvider 제거 및 JwtProvider로 대체
- MemberRepository 대신 MemberRepositoryFacade 도입 및 의존성 주입 방식 수정
- 테스트에서 mock 사용 제거 및 실제 데이터 기반 테스트로 전환

changuii committed:
feat: ReservationRepositoryFacade에 새로운 조회 및 존재 여부 확인 메서드 추가

ReservationRepositoryFacade에 `findAllByThemeAndDate` 및 `existsByReservationTime` 메서드를 추가하고,
ReservationTimeService에서 이를 활용하도록 변경했습니다.

changuii committed:
feat: JwtProvider 설정 외부화 및 application.yml에 비밀키/유효기간 설정 추가

JwtProvider의 비밀키와 유효기간을 application.yml로 외부화하여 관리하도록 변경. @Value를 통해 주입받고, @PostConstruct를 통해 Key 초기화 로직 추가.

changuii committed:
test: 불필요한 MockitoExtension 어노테이션 제거 및 테스트 설정(application.yml) 수정

테스트 클래스에서 사용하지 않는 @ExtendWith(MockitoExtension.class) 어노테이션 제거.
테스트 설정(application.yml)에 JWT 키 및 만료시간 추가.

changuii commented on src/main/java/roomescape/member/MemberRepositoryFacadeImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final MemberRepository memberRepository;
+
+    public void save(Member member) {
+        memberRepository.save(member);
+    }
```
내 생각에는 지킬 수 있는 곳에는 CQS를 지키는게 더 좋은 것 같다고 판단했어

changuii replied:
@coderabbitai review

changuii committed:
test: AuthServiceTest에 passwordEncoder 추가하여 비밀번호 인코딩 적용

changuii commented on src/test/java/roomescape/theme/ThemeServiceTest.java:
```diff
@@ -65,18 +101,22 @@ class FindAll {
+                    new Theme("로키1", "로키로키1", "http://www.google.com/1"),
+                    new Theme("로키2", "로키로키2", "http://www.google.com/2"),
+                    new Theme("로키3", "로키로키3", "http://www.google.com/3")
             );
-            given(themeRepository.findAll())
-                    .willReturn(themes);
+            for (final Theme theme : themes) {
+                themeRepositoryFacade.save(theme);
+            }
 
             // when
             final List<ThemeResponse> actual = themeService.findAll();
 
             // then
-            assertThat(actual).hasSize(3);
+            assertThat(actual)
+                    .hasSize(3)
+                    .extracting(ThemeResponse::id)
+                    .contains(1L, 2L, 3L);
         }
```
Rank에 정렬기준에 맞게 작성해뒀어 만약, 정렬 순서가 변경된다면 기능 변경이기 때문에 테스트를 수정해주는게 맞다고 생각해

현재 구조라면 정렬 기준이라는 기능의 변경 없이 리팩터링된다면 테스트는 깨지지 않아

@coderabbitai review

changuii commented on src/test/java/roomescape/theme/ThemeServiceTest.java:
```diff
@@ -43,16 +80,15 @@ class Create {
-            final Theme theme = new Theme(themeRequest.name(), themeRequest.description(), themeRequest.thumbnail());
-            final Theme savedTheme = new Theme(1L, theme.getName(), theme.getDescription(), theme.getThumbnail());
-            given(themeRepository.save(theme))
-                    .willReturn(savedTheme);
+            final Theme theme = new Theme(1L, themeRequest.name(), themeRequest.description(),
+                    themeRequest.thumbnail());
+            final ThemeResponse expected = ThemeResponse.from(theme);
 
```
너가 말한 부분은 expected, 즉 기댓값을 만들기 위한 변수야
- 그렇기 때문에 내가 기대하는 id를 작성한거야

@coderabbitai review

changuii commented on src/test/java/roomescape/theme/ThemeServiceTest.java:
```diff
@@ -2,36 +2,73 @@
     private final ThemeService themeService;
-    private final ReservationRepository reservationRepository;
 
-    public ThemeServiceTest() {
-        themeRepository = mock(ThemeRepository.class);
-        reservationRepository = mock(ReservationRepository.class);
-        themeService = new ThemeService(themeRepository, reservationRepository);
+    private final ReservationTimeRepositoryFacade reservationTimeRepositoryFacade;
+    private final MemberRepositoryFacade memberRepositoryFacade;
+    private final ReservationRepositoryFacade reservationRepositoryFacade;
+
```
난 데이터베이스에 저장, 삭제까지가 하나의 기능으로 판단했고 단위 테스트라고 생각하고 코드를 작성했어

특히, 너가 말한 부분은 command에 해당하는 메서드로 delete의 반환값은 void이기 때문에 외부로 나가는 요청을 검증한거야
- 왜냐하면 내가 기대하는 delete의 결과값은 반환값이 아닌, delete 메서드의 호출이기 때문이야

https://github.com/coderabbitai review

changuii commented on src/test/java/roomescape/reservationtime/ReservationTimeServiceTest.java:
```diff
@@ -53,17 +85,13 @@ void createTime1() {
-            given(reservationTimeRepository.save(new ReservationTime(startAt)))
-                    .willReturn(reservationTime);
+            final ReservationTimeResponse expected = ReservationTimeResponse.from(new ReservationTime(1L, startAt));
 
             // when
             final ReservationTimeResponse actual = reservationTimeService.create(request);
 
             // then
-            assertThat(actual).isEqualTo(ReservationTimeResponse.from(reservationTime));
+            assertThat(actual).isEqualTo(expected);
         }
```
각 단위 테스트간에는 increment id와 테이블을 초기화해서 격리시켜줬어
- 너가 말한 id가 존재하는 코드는 기댓값이야 따라서 내가 기대하는 id 값을 하드코딩한거야

@coderabbitai review

changuii commented on src/test/java/roomescape/reservationtime/ReservationTimeServiceTest.java:
```diff
@@ -2,45 +2,77 @@
+})
 public class ReservationTimeServiceTest {
 
+    @MockitoSpyBean
+    private final ReservationTimeRepositoryFacade reservationTimeRepositoryFacade;
     private final ReservationTimeService reservationTimeService;
```
너가 잘못 알고 있는거야
MockitoSpyBean은 applicationContext에 존재하는 빈을 spy로 감싸 주입해주는게 아니라, applicationContext에 존재하는 해당 빈을 spy로 감싸기만하고 실제 주입은 Autowired가 해주기 때문에 그런 문제는 유발하지 않아

@coderabbitai review

changuii committed:
feat: 예약 대기 기능 추가

- 예약 페이지에 '예약대기' 버튼 추가 및 비활성화/활성화 로직 구현
- 'onWaitButtonClick' 함수 추가로 예약 대기 요청 처리 로직 작성
- application.yml의 jwt-expiration 값을 밀리초로 수정
- 사용하지 않는 코드를 정리 및 예외 로그 출력 위치 수정

changuii committed:
fix: 내 예약 목록 조회 API endpoint 수정

changuii committed:
feat: 예약 대기 기능 구현

- 예약 대기 생성 API 추가
- ReservationStatus에 WAITING 상태 추가
- 예약 중복 및 상태 검증 로직 확장
- 선택적 reservation 대기 처리 로직 도입
- 기존 reservation 관련 테스트 데이터 정리 및 추가

changuii committed:
test: reservation waiting 생성 테스트 추가

reservation waiting 생성 로직에 대한 정상 시나리오 및 예외 케이스 (존재하지 않는 멤버, 예약 시간, 테마, 중복 예약, 과거 날짜/시간 예약 등)를 검증하는 테스트를 추가함.

changuii committed:
feat: 예약 대기 관리 페이지 추가

예약 대기 목록 조회와 관련된 admin/waiting 엔드포인트, HTML 뷰, JS 스크립트를 추가하고, 테스트 코드를 작성하여 페이지 요청 시 정상 동작을 확인합니다.

changuii committed:
feat(reservation): 예약 대기 관리 기능 구현

예약 대기 목록 조회 API 및 관련 엔드포인트 추가
- "/waiting" GET 엔드포인트 구현
- 예약 상태에 따른 목록 필터링 및 처리 기능 추가
- 예약 삭제 시 다음 대기 예약자 대기 상태로 전환 처리 로직 추가

changuii committed:
test: 예약 삭제 시 대기 중 예약 상태 변경 테스트 추가

changuii committed:
feat: 예약 대기 생성 및 조회 기능 추가

예약 대기 생성 기능과 모든 예약 대기 조회 기능을 구현했습니다.
- 대기 상태의 예약 랭킹을 계산하여 반환하기 위한 `WaitingRankReservation` 클래스 추가
- 예약 대기가 존재하지 않을 경우 예외 처리 `ReservationNotExistsPendingException` 추가
- `/reservations/waiting` API로 예약 대기를 생성하고 조회할 수 있도록 컨트롤러, 서비스, 레포지토리 수정
- 관련 테스트 추가 및 `ReservationRepository`에 JPQL 기반 쿼리 추가

changuii commented:
@coderabbitai full review

Hyunta commented on src/main/java/roomescape/member/MemberRepositoryFacadeImpl.java:
```diff
@@ -0,0 +1,37 @@
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Repository;
+
+@Repository
+@AllArgsConstructor
+public class MemberRepositoryFacadeImpl implements MemberRepositoryFacade{
+
+    private final MemberRepository memberRepository;
```
RepositoryFacade는 왜 필요한가요?

changuii replied:
RepositoryFacade를 추가한 이유는, service는 jpa와 같은 데이터 접근 기술을 모르게 하고 싶었습니다.
jpa를 도입하더라도 여전히 다른 데이터 접근 기술을 사용할 여지가 있었습니다. 추가로 jpa와 다른 기술을 같이 사용해서 만들수도 있을 것 같습니다.
이전에는, jpaRepository를 바로 사용하여 변경이 어려웠다면 현재는 facade의 추가로, service의 변경없이 jdbc로 변경할 수 있습니다.

Hyunta replied:
이전에도 말씀 드렸지만 우선 데이터 접근 기술이 바뀌는 경우는 극히 드물게 일어납니다.
반대로 데이터를 가져오는 메서드를 만드는 경우는 훨씬 잦겠죠. SpringDataJPA를 사용하면서 가져가는 생산성의 절반 이상을 잃어버린다고 생각합니다. Facade 라는 레이어를 추가했지만 또 막상 Facade에 들어가면 그대로 Repository의 메서드를 호출하고 있습니다. 여러모로 유지하는데 번거로운데에 반해 가져가는 이점이 데이터 접근 기술에 대한 대비라고 하기에는 잘 일어나지도 않을 뿐더러 막상 변경을 해야할 때 공수를 산정하더라도 그렇게 손이 많이 가는 일은 아닙니다.

혹시 일어날 일을 대비해서 확장성을 가져가자고 주장할 때는 저는 발생 가능성과 후작업에 대한 난이도를 생각해봐야 한다고 생각하는데, 발생 가능성도 적고, 후작업에 대한 난이도도 그렇게 높지 않아서 지나친 확장에 대한 고려라고 보여집니다.

changuii replied:
제 개인적인 생각으로 JpaRepository와 Service 사이에 새로운 계층을 추가하고, 새로운 쿼리가 생길때마다 한번더 위임하는 메서드를 만드는게 그렇게 번거로운 작업은 아니었습니다.
- 아서가 말씀하신대로 계층이 하나 추가되어 위임하는 메서드를 하나 더 작성하는게 Jpa를 사용하면서 가져가는 생산성의 절반 이상을 잃어버리는 것은 동의하기 힘들다고 생각합니다.
	- 제가 생각하는 Jpa를 사용하면서 발생하는 생산성 이점은 다음과 같습니다.
		- 쿼리를 작성하지 않더라도 대부분의 쿼리를 SpringDataJPA가 만들어주는 것
		- JDBC를 사용할때와는 달리, 우리가 따로 RowMapper를 일일이 구현하지 않아도 되는 것
	- 아서가 말씀하신 이점은 JDBC를 인터페이스로 바로 사용하더라도 가져갈 수 있는 이점이라 생각합니다. 이게 JPA를 사용함으로써 얻는 이점이라기 보다는 직접적으로 데이터 접근 기술에 의존해서 편리하게 사용하는 것에서 발생하는 이점이라고 생각합니다!

하지만, 제가 느낀 경험은 만약 JpaRepository가 없어지는 상황에서 이런 중간 인터페이스가 존재하지 않는다면 어떤 메서드를 사용중인지 바로 파악하기가 어려웠습니다.(CrudRepository에 명세된 메서드들) 
특히, 이런 repository가 없다면 service와 repository를 분리하여 테스트를 수행하려고 할때 분리하기가 어렵습니다.
- FakeRepository를 만드려고 할때, spring이 만들고 있는JpaRepository의 사용하지 않는 모든 메서드들을 구현해주어야 하는데 번거로운 작업이 될 수 있습니다.

따라서 제가 생각했을 때는 현재 중간 계층을 추가함으로써 얻는 불편함이 미래의 확장성으로 인한 이점보다 작다고 판단했고 이런 판단이 지나친 확장성에 대한 고려라고 판단하지 않았습니다.

Hyunta replied:
네 경험해보는 것은 좋다고 생각합니다.
우선 이렇게 유지하면서 장단점을 느껴보시면 좋을 것 같습니다👍🏻

Hyunta commented on src/main/java/roomescape/reservation/ReservationRepositoryFacade.java:
```diff
@@ -0,0 +1,42 @@
+    Optional<Reservation> findById(Long id);
+
+    List<Reservation> findAll();
+    List<Reservation> findAllByThemeAndDate(Theme theme, LocalDate date);
+    List<Reservation> findAllByMemberAndThemeAndDateBetween(Member member, Theme theme, LocalDate from, LocalDate to);
+    List<Reservation> findAllByReservationStatus(ReservationStatus reservationStatus);
+    List<Reservation> findAllByDateAndReservationTimeAndThemeAndReservationStatusOrderByAsc(
+            LocalDate date,
+            ReservationTime reservationTime,
+            Theme theme,
```
Facade를 만들면서 SpringDataJPA에서 가져가는 이점이 사라진게 아닐까요?

changuii replied:
Spring Data Jpa를 사용하여 편리하게 쿼리들을 추가할 수 있는 장점이 facade의 도입으로 인해 불편하게 느껴질 수 있지만, 위임 메서드 하나만 추가하면 되기 때문에 단점이 그리 크다고 생각하지 않았습니다.
facade를 도입함으로써 데이터 접근 기술을 쉽게 변경할 수 있는 이점이 더 크다고 생각하였습니다.

Hyunta commented on src/main/java/roomescape/reservation/ReservationRepositoryFacadeImpl.java:
```diff
@@ -0,0 +1,93 @@
+import roomescape.reservationtime.ReservationTime;
+import roomescape.theme.Theme;
+
+@Repository
+@RequiredArgsConstructor
+public class ReservationRepositoryFacadeImpl implements ReservationRepositoryFacade {
```
`@Repository` 애노테이션이 왜 필요한가요?

changuii replied:
Repository 어노테이션을 붙여서 빈으로 관리함과 동시에 이 클래스가 repository와 관련있음을 명시했습니다.

Hyunta commented on src/test/java/roomescape/auth/AuthServiceTest.java:
```diff
@@ -2,35 +2,48 @@
 import roomescape.member.MemberRole;
 
-@ExtendWith(MockitoExtension.class)
+@DataJpaTest
+@Sql(scripts = "classpath:/initialize_database.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Import({
```
테스트 시 데이터 초기화 하는 방법을 어떻게 선정하셨을까요?

changuii replied:
데이터 초기화 방법에 대해서 다음의 요소들을 고려하였습니다.
1. JdbcTemplate로 삭제 쿼리
2. sql 스크립트 사용
3. DirtiesContext로 스프링 컨텍스트 초기화 

JdbcTemplate는 클래스간에 공유해서 초기화 로직을 수행할 수 없었고 매번 클래스에 초기화를 해주어야 하는 단점이 존재했습니다.
DirtiesContext는 데이터를 초기화해줄 뿐만 아니라 스프링 컨텍스트 빈들도 다시 로드되기 때문에 불필요하게 빈을 초기화할 필요는 없다고 판단하여 사용하지 않았습니다. 

그래서 여러 클래스간에 공유해서 사용할 수 있고, 데이터만 제거할 수 있는 sql 스크립트를 사용하였습니다.

Hyunta commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -55,13 +58,38 @@ public ReservationResponse createForAdmin(final AdminReservationRequest request)
+        final Theme theme = getThemeById(request.themeId());
+        final Member member = getMemberByEmail(loginMember.email());
+
+        validateNotExistsPendingReservation(request.date(), reservationTime, theme);
+        validateDuplicateReservation(request, reservationTime, theme, member);
+        validatePastDateTime(request.date(), reservationTime);
```
검증하는 로직을 객체에 위임하기는 어려웠을까요?

changuii replied:
죄송합니다. 아서!
제가 의도가 무엇인지 잘 파악이 안됩니다.

1. Reservation 도메인 객체 내부에 검증 로직을 위임하라는 의미인가요?
	- 최대한 객체 내부로 넣을 수 있는 검증 로직이라면 넣어주었고, 쿼리가 필요한 검증이라면 넣어주지 않았습니다.!
2. 아니면, validator와 같은 검증을 위한 객체를 만들어 위임하라는 의미인가요?
	- validator를 두는 큰 의미를 못느꼈고, service에서 로직의 흐름을 모두 읽을 수 있는 것이 가독성이 더 높을 것 같습니다.

Hyunta replied:
전자의 의도로 드린 질문이었습니다.
전부 Reservation안에서 검증할 수 없는 로직들인가요?

changuii replied:
시간과 날짜의 경우에는 내부에서 검증할 수 있습니다... 죄송합니다! 아서
ReservationDate VO를 만들어 과거 시간으로 생성할 수 없도록 검증하는 책임을 가지도록 만들었습니다.

이 지점에서 고민이 생겼습니다.
Reservation을 생성하는 경우에는 과거의 날짜로 예약을 해서는 안되기 때문에 이를 검증해서 생성해야 하지만 ReservationDate로 조회를 하는 경우에는 과거 날짜로 조회할 수 있어야 했습니다.
정적 팩터리 메서드를 통해서 검증을 회피할 수 있도록 만들었으나 네이밍을 통해서 이 의도를 잘 전달할 수 있는지 잘 모르겠습니다.

Hyunta replied:
검증하는 로직을 한번에 관리하는 것이 중요하다고 생각해서 드린 커멘트였습니다.
검증하는 로직을 여러 곳에서 매번 생성하지 않고 한번에 관리할 수 있으면 그대로 충분하다고 생각합니다.

Hyunta commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -103,30 +157,52 @@ private void validatePastTime(final ReservationTime reservationTime) {
+            final LocalDate date, final ReservationTime reservationTime,
+            final Theme theme
+    ) {
+        if (reservationRepositoryFacade.existsByReservationTimeAndDateAndThemeAndReservationStatus(
```
파사드를 사용할거라면 네이밍을 조금 줄여도되지 않을까요?

changuii replied:
저도 쿼리 메서드가 너무 길어지면서 불편하다고는 생각했지만 현재 네이밍이 jpa를 알고 있다면 읽는 사람이 가장 직관적으로 이해할 수 있을 것이라 생각하여 그대로 뒀습니다.
- 다만, 더 긴 쿼리들이 나올 수 있다는 점에서 코드가 읽기 힘들어져 어떤 방법을 하면 좋을지 고민해보았습니다.

어떤 방법으로 줄일 수 있을까?
1. 도메인 용어 활용 -> DuplicateReservation: 현재 쿼리는 reservation이 중복되어서는 안되는 조건으로 쿼리를 보내고 있기 때문에, 나중에 추가되더라도 자연스럽다.
2. existsBy로 통일 -> 매개변수 타입만으로 충분한 정보를 제공하기 어렵고, 어떤 시그니처들이 존재하는지 찾아보아야 하기 때문에 직관성이 떨어진다.

existsBy로 메서드들을 오버로딩하는 것보다는 쉽게 이해가능한 existsByDuplicateMember로 변경하였습니다.

Hyunta commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -70,23 +98,49 @@ public List<ReservationResponse> readAllByMemberAndThemeAndDateRange(final Admin
+        if (!waitingReservations.isEmpty()) {
+            final Reservation nextReservation = waitingReservations.getFirst();
+            nextReservation.pending();
+            reservationRepositoryFacade.save(nextReservation);
```
save를 생략해도 되지 않나요?

changuii replied:
save를 생략하더라도 제가 transactional 어노테이션을 추가해주었기 때문에 제가 따로 save()를 호출하지 않더라도 메서드가 종료되면서 커밋 시점에 update와 delete 데이터베이스 쿼리로 날라갑니다.
저 시점에 명시적으로 save()를 통해 update를 하지 않더라도 괜찮기 때문에 반영하였습니다!

Hyunta reviewed (changes requested):
안녕하세요 부기,
이번 미션도 잘 구현해주셨네요.
몇가지 커멘트 남겼으니 확인부탁드립니다.

changuii committed:
fix: open-in-view 설정 false로 변경

changuii committed:
style(test): @Import 배열 포매팅 수정 및 가독성 개선

changuii committed:
fix: 예약 대기 상태 변경 시 불필요한 저장 로직 제거

changuii committed:
refactor(reservation): 예약 중복 검증 메서드 리팩토링 및 명확한 메서드명으로 변경

중복 검증 메서드 및 관련 Repository 메서드명을 명확하게 수정하였고, 유사 로직을 통합하여 코드 중복을 제거했습니다.
- validateDuplicateDateTimeAndTheme → validateDuplicatePending으로 변경
- validateNotExistsPendingReservation → validateNotExistsPending으로 변경
- validateDuplicateReservation → validateDuplicateMember로 통합
- Repository 메서드명 변경: existsByDateAndReservationTimeAndThemeAndMember → existsByDuplicateMember, existsByReservationTimeAndDateAndThemeAndReservationStatus → existsDuplicateStatus

changuii committed:
refactor: 메서드명 및 변수명 수정으로 코드 가독성 향상

ThemeService와 ReservationTimeService의 메서드명과 관련된 변수명을 보다 명확하고 간결하게 수정하여 코드의 가독성을 개선했습니다. (e.g., `findTopRankThemes` -> `findTopRank`, `validateDuplicateTime` -> `validateDuplicate` 등)

changuii committed:
refactor: 기존, Repository를 JpaRepository로, RepositoryFacade를 Repository로 변경

Hyunta reviewed (changes requested):
안녕하세요 부기,
피드백 반영도 빠르게 잘해주셨네요.
추가로 커멘트 남겨드렸는데 확인 후 재요청주세요.
궁금하시거나 모호한 부분이 있으면 DM 주세요.

changuii committed:
fix(auth): 로그인 쿠키 설정 강화 (Secure, HttpOnly, SameSite 추가)

changuii committed:
chore: open-in-view 설정 제거

changuii committed:
perf: Reservation 엔티티에 Lazy 로딩 적용 및 JPQL fetch join 추가

Reservation 엔티티에 @ManyToOne 관계에 대해 Lazy 로딩을 설정하여 성능 최적화.
ReservationJpaRepository에 fetch join을 활용한 JPQL 쿼리 추가로 데이터 조회 효율성 강화.

changuii committed:
feat(reservation): 예약 조회 시 연관 엔티티(theme, member, reservationTime) join fetch 추가

changuii committed:
refactor(reservation): ReservationDate 를 추가하여 날짜 및 시간 로직 개선

ReservationDate 클래스를 도입하여 명시적으로 예약 날짜와 시간을 다루도록 리팩토링했습니다. 이에 따라 관련 메서드 및 테스트 코드에서 LocalDate를 ReservationDate로 변경하고 가독성과 사용성을 향상시켰습니다.

changuii commented:
안녕하세요 아서! 리뷰 감사합니다. :)

### 추가한 작업
- 토큰 정보를 더 안전하게 보관하는 기능 추가 (secure, httpOnly, SameSite)
- 자주 조회하는 내용에 대해 LAZY로 변경하고 필요한 부분만 fetch join 적용
- reservation date를 VO로 감싸 책임 부여

changuii committed:
refactor: 도메인 분리를 위한 패키지 구조 리팩토링

- `reservation`, `reservationtime`, `theme` 등의 도메인 엔티티 및 관련 클래스를 각각의 하위 패키지 (`domain`, `repository`, `dto`)로 이동
- 명확한 계층 구조를 반영하여 코드 가독성과 유지보수성 향상

changuii committed:
refactor(reservation): WaitingRankReservation 파일 위치 변경 및 관련 경로 수정

WaitingRankReservation 클래스의 패키지를 dto에서 domain으로 이동하고, 관련된 클래스 및 쿼리 경로를 수정함.

changuii committed:
test: Reservation 및 ReservationDate 테스트 코드 추가

Reservation 및 ReservationDate 도메인에 대한 테스트 코드를 작성하여 검증을 강화했습니다.
- `ReservationDateTest` 작성: 날짜 비교 및 예외 처리 테스트
- `ReservationTest` 작성: 예약 상태 및 시간 검증 테스트
- `isToday` 메서드를 `isEqualToDate`로 변경하여 의미를 명확히 수정

Hyunta reviewed (approved):
안녕하세요 부기,
남겨주신 커멘트에 답변 달아드렸습니다.
추가로 궁금하시거나 모호한 부분이 있으면 DM 주세요.
이번 단계 진행하시느라 고생하셨습니다!
